yapapi/services.py

```diff
@@ -191,26 +204,143 @@ def _respond_nowait(self, message: Optional[Any], response_to: Optional[ServiceS
+        service's instance of the work context) that are then dispatched to the activity by the
+        engine.
+
+        Results of those batches can then be captured by awaiting the values captured from yield
+        statements.
+
+        ### Example
+
+        ```
+        async def start(self):
+            # deploy the exe-unit
+            self._ctx.deploy()
+            # start the exe-unit's container
+            self._ctx.start()
+            # start some service process within the container
+            self._ctx.run("/golem/run/service_ctl", "--start")
+            # send the batch to the provider
+            yield self._ctx.commit()
+        ```
+
+        ### Default implementation
+
+        The default implementation assumes that, in order to accept commands, the runtime needs to
+        be first deployed using the `deploy` command, which is analogous to creation of a container
+        corresponding with the desired payload, and then started using the `start` command,
+        actually launching the process that runs the aforementioned container.
+
+        Additionally, it also assumes that the exe-unit doesn't need any additional parameters
+        in its `start()` call (e.g. for the VM runtime, all the required parameters are already
+        passed as part of the agreement between the requestor and the provider).
+
+        Therefore, this default implementation performs the minimum required for a VM payload to
+        start responding to `run` commands. If your service requires any additional operations -
+        you'll need to override this method (possibly starting with a call to `super().start()`)
+        to add appropriate preparatory steps.
+
+        In case of runtimes other than VM, `deploy` and/or `start` might be optional or altogether
+        disallowed, plus `start` itself might take some parameters. It is up to the author of the
+        specific `Service` implementation that uses such a payload to adjust this method accordingly
+        based on the requirements for the given runtime/exe-unit type.
+        """
 
         self._ctx.deploy()
         self._ctx.start()
         yield self._ctx.commit()
 
-    async def run(self):
-        """Implement the `running` state of the service."""
+    async def run(self) -> AsyncGenerator[WorkItem, Awaitable[List[events.CommandEvent]]]:
+        """Implement the handler for the `running` state of the service.
+
+        To be overridden by the author of a specific Service class.
+
+        Should contain any operations needed ensure the continual operation of a service.
+
+        As a handler, it's expected to be a generator that yields `WorkItems` (generated using the
+        service's instance of the work context) that are then dispatched to the activity by the
+        engine.
+
+        Results of those batches can then be captured by awaiting the values captured from yield
+        statements.
+
```

Comment:
I would add here an explicit note that when this function exits the service will be stopping.
I understand this is a more-or-less direct consequence of the default implementation, but might not be obvious enough.

Reply:
updated